content/advent-2017/lexmachine-advent.md

```diff
@@ -0,0 +1,644 @@
++++
+author = ["Tim Henderson"]
+title = "How to Tokenize Complex Strings with Lexmachine"
+linktitle = "How to Tokenize Complex Strings with Lexmachine"
+date = 2017-12-12
+draft = false
++++
+
+
+This article is about using [lexmachine](https://github.com/timtadh/lexmachine)
+to tokenize strings (split up into component parts) in the Go (golang)
+programming language. If you find yourself processing a complex file format or
+network protocol this article will walk you through how to use `lexmachine` to
+process both accurately and quickly. If you need more help after reading this
+article take a look at the
+[documentation](https://github.com/timtadh/lexmachine#documentation), a
+[tutorial](http://hackthology.com/writing-a-lexer-in-go-with-lexmachine.html),
+or an [explainer
+article](http://hackthology.com/faster-tokenization-with-a-dfa-backend-for-lexmachine.html)
+on how it all works.
+
+# What is string tokenization?
+
+String tokenization takes a piece of text and splits it into categorized
+substrings. When conducted as part of parsing or compiling a programming
+language it may also be called *lexical analysis*. Depending on the kind of text
+being processed splitting it up might be very simple or extremely complicated.
+For instance, some simple text formats (such as basic CSV or TSV files) separate
+each record by newline characters (`\n` or ASCII `0x0A`). At the other end of
+the spectrum, natural language (such as English text) may be very difficult to
+correctly tokenize.
+
+Let's look at a quick example of tokenizing an English sentence. 
+
+
+> Mary had a little lamb and its fleece was white as snow.
+
+Each word will be a separate token.  A *token* is a pair made up of the *lexeme*
+(or substring) and the *token type* (or category). When tokenizing English,
+often the token type will be the part of speech.
+
+```
+ Mary           had     a          little      lamb    and            ...
+<Proper Noun>  <Verb>  <Article>  <Adjetive>  <Noun>  <Conjunction>   ...
+```
+
+Tokenizing English is very difficult in general because determining the part of
+speech of a particular word is not always obvious. The same word can play
+multiple *semantic* roles in a sentence. Humans to consider the sentence in its
```

Comment:
Speaking of lexical analysis, does the sentence starting with `Humans` need a verb, or did I improperly tokenize it?  :D

Reply:
fixed.